projects/optic/src/commands/diff/diff-all.ts

```diff
@@ -0,0 +1,452 @@
+  }
+  return results;
+}
+
+async function computeAll(
+  candidatesMap: ReturnType<typeof matchCandidates>,
+  config: OpticCliConfig,
+  options: DiffAllActionOptions
+): Promise<{
+  warnings: Warnings;
+  results: Result[];
+}> {
+  const warnings: Warnings = {
+    missingOpticUrl: [],
+    unparseableFromSpec: [],
+    unparseableToSpec: [],
+  };
+
+  const results: Result[] = [];
+
+  for await (const [_, candidate] of candidatesMap) {
+    // try load both from + to spec
+    let fromParseResults: ParseResult;
+    let toParseResults: ParseResult;
+    try {
+      fromParseResults = await getFileFromFsOrGit(
+        candidate.from,
+        config,
+        false
+      );
+    } catch (e) {
+      warnings.unparseableFromSpec.push({
+        path: candidate.from!,
+        error: e,
+      });
+      continue;
+    }
+
+    try {
+      toParseResults = await getFileFromFsOrGit(candidate.to, config, true);
+    } catch (e) {
+      warnings.unparseableToSpec.push({
+        path: candidate.to!,
+        error: e,
+      });
+      continue;
+    }
+
+    // Cases we run the comparison:
+    // - if to spec has x-optic-url
+    // - if from spec has x-optic-url AND to spec is empty
+    if (
+      typeof toParseResults.jsonLike[OPTIC_URL_KEY] === 'string' ||
+      (typeof fromParseResults.jsonLike[OPTIC_URL_KEY] === 'string' &&
+        toParseResults.isEmptySpec)
+    ) {
+      logger.info(
+        chalk.blue(
+          `Diffing ${candidate.from ?? 'empty spec'} to ${
+            candidate.to ?? 'empty spec'
+          }`
+        )
+      );
+      const { specResults, checks, changelogData } = await compute(
+        [fromParseResults, toParseResults],
+        config,
+        options
+      );
+
+      if (specResults.changes.length === 0) {
+        logger.info('No changes were detected');
+      }
+      logger.info('');
+
+      for (const log of terminalChangelog(changelogData)) {
+        logger.info(log);
+      }
+
+      if (options.check) {
+        if (specResults.results.length > 0) {
+          logger.info('Checks');
+          logger.info('');
+        }
+
+        for (const log of generateComparisonLogs(specResults, {
+          output: 'pretty',
+          verbose: false,
+        })) {
+          logger.info(log);
+        }
+
+        logger.info('');
+      }
+
+      results.push({
+        fromParseResults,
+        toParseResults,
+        specResults,
+        checks,
+        changelogData,
+        from: candidate.from,
+        to: candidate.to,
+      });
+    } else if (
+      !toParseResults.isEmptySpec &&
+      typeof toParseResults.jsonLike[OPTIC_URL_KEY] !== 'string'
+    ) {
+      warnings.missingOpticUrl.push({
+        path: candidate.to!,
+      });
+      continue;
+    }
+  }
+  return {
+    warnings,
+    results,
+  };
+}
+
+type Result = Awaited<ReturnType<typeof compute>> & {
+  fromParseResults: ParseResult;
+  toParseResults: ParseResult;
+  from?: string;
+  to?: string;
+};
```

Comment:
I'm not a fan of this results format, it's pretty massive given it includes the entire specs + results, which can mean we are using a ton of memory. I originally had this inside the `compute` function, but moved it out for code simplicity (to handle JSON / terminal output diffs computed from results)

I think this is _ok_ for now, since this is how `bulk-compare` works, but I'm pretty sure we'll run into issues with a large number of diff comparisons

This will be made better when we have the new results format which will drastically reduce the memory usage, however keeping all files in memory will keep the memory usage high

Thoughts?

Reply:
I think it's fine until it's not